src/cpp/fastdds/rpc/ReplierImpl.cpp

```diff
@@ -302,22 +326,30 @@ ReplierImpl::RequesterMatchStatus ReplierImpl::requester_match_status(
         return RequesterMatchStatus::UNMATCHED;
     }
 
-    FASTDDS_TODO_BEFORE(3, 3, "Get reply reader GUID from pub_data");
-
     auto related_guid = info.related_sample_identity.writer_guid();
     bool reply_topic_matched = false;
     if (info.sample_identity.writer_guid() != related_guid)
     {
         // Custom related GUID (i.e. reply reader GUID) sent with the request.
         // Check if the replier writer is matched with that specific reader.
-        SubscriptionBuiltinTopicData sub_data;
-        reply_topic_matched = RETCODE_OK == replier_writer_->get_matched_subscription_data(sub_data, related_guid);
+        if (related_guid.entityId.is_reader())
+        {
+            SubscriptionBuiltinTopicData sub_data;
+            reply_topic_matched = RETCODE_OK == replier_writer_->get_matched_subscription_data(sub_data, related_guid);
+        }
+        else
+        {
+            // At least, demand that the number of matched entities on the replier
+            // writer is equal to the ones on the requester reader.
+            reply_topic_matched = is_fully_matched();
```

Comment:
This should use the `related_datareader_key`. So I would change line 331 to `if (related_guid.entityId.is_reader() && (info.sample_identity.writer_guid() != related_guid))` and always fallback to the code below in the `else`